Implement environment call properly

Risheng1128 opened the pull request:
Implement ecall instruction with control and status registers (CSRs) 
to pass the riscv-arch-test with all privilege tests.

riscv-arch-test result:
```shell
Check ebreak                    ... OK 
Check ecall                     ... OK 
Check misalign1-jalr-01         ... OK 
Check misalign2-jalr-01         ... OK 
Check misalign-beq-01           ... OK 
Check misalign-bge-01           ... OK 
Check misalign-bgeu-01          ... OK 
Check misalign-blt-01           ... OK 
Check misalign-bltu-01          ... OK 
Check misalign-bne-01           ... OK 
Check misalign-jal-01           ... OK 
Check misalign-lh-01            ... OK 
Check misalign-lhu-01           ... OK 
Check misalign-lw-01            ... OK 
Check misalign-sh-01            ... OK 
Check misalign-sw-01            ... OK
```

jserv reviewed (changes requested):
I failed to run `make check` because of no progress.

Risheng1128 committed:
Implement environment call properly

Implement ecall instruction with control and status
registers (CSRs).

However, there may occur the situation that the elf
file doesn't contain any csr instruction. It will make
pc point to wrong address.

To resolve this, using 0 as the initial value of mtvec.
By default, the emulator executes default exception
handler when an exception occurs and mtvec stores 0.

jserv commented on src/emulate.c:
```diff
@@ -61,9 +72,15 @@ enum {
         rv->csr_mepc = rv->PC;                                              \
         rv->csr_mtval = mtval;                                              \
+        rv->csr_mcause = code;                                              \
+        if (!rv->csr_mtvec) { /* elf doesn't implement csr instruction */   \
```
The comment was misleading. You can say,
> in case CSR is not configured.

jserv commented on src/emulate.c:
```diff
@@ -40,14 +40,25 @@ extern struct target_ops gdbstub_ops;
+    _(store_fault)      /* Store/AMO access fault */         \
+    _(ecall_U)          /* Environment call from U-mode */   \
+    _(ecall_S)          /* Environment call from S-mode */   \
+    _(reserved)         /* Reserved */                       \
```
Shall we care about "Reserved" item?

Risheng1128 replied:
I think the "reserved" item should be removed. Maybe I need to
modify the exception list with assigning exception code and remove
unuseful items.

jserv replied:
In addition, check ecall_{U,S} as well since only `ecall_M` makes sense now.
